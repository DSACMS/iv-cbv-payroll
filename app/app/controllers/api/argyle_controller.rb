class Api::ArgyleController < ApplicationController
  before_action :set_cbv_flow, :resume_previous_argyle_account_connection
  after_action :track_event

  # This API endpoint is used to fetch a `user_token` to allow the user to open
  # the Argyle modal.
  #
  # @see https://docs.argyle.com/link/user-tokens
  def create
    is_sandbox_environment = agency_config[@cbv_flow.client_agency_id].argyle_environment == "sandbox"
    user_token = if @cbv_flow.argyle_user_id.blank?
                   response = argyle.create_user(@cbv_flow.end_user_id)

                   # Store the argyle_user_id to allow us to associate incoming webhooks with
                   # this CbvFlow.
                   @cbv_flow.update(argyle_user_id: response["id"])

                   response["user_token"]
                 else
                   # If the user has already been created in Argyle, let's just
                   # make them a new link token with the same user.
                   response = argyle.create_user_token(@cbv_flow.argyle_user_id)
                   response["user_token"]
                 end

    render json: {
      status: :ok,
      isSandbox: is_sandbox_environment,
      flowId: Aggregators::Sdk::ArgyleService::FLOW_ID,
      user: {
        user_token: user_token
      }
    }
  end

  private

  def set_cbv_flow
    @cbv_flow = CbvFlow.find_by(id: session[:cbv_flow_id])
    redirect_to(root_url(cbv_flow_timeout: true)) unless @cbv_flow
  end

  # Redirect if the user is attempting connect a previously connected account
  def resume_previous_argyle_account_connection
    item_id = params[:item_id]
    unless item_id.present?
      return render json: { status: :error, message: "Invalid item_id" }, status: :unprocessable_entity
    end

    return if item_id == "item_000026933" # If it's an ADP item, allow user to create a new connection.

    # If the user ID is not yet set, there is no previous argyle session to resume
    return unless @cbv_flow.argyle_user_id.present?

    # Find any previous Argyle connections to this item
    connected_argyle_accounts = argyle.fetch_accounts_api(user: @cbv_flow.argyle_user_id, item: item_id)["results"]
    return unless connected_argyle_accounts.any?

    # Find the PayrollAccount object (if we have received an accounts.connected webhook)
    argyle_account = connected_argyle_accounts.first
    payroll_account = @cbv_flow.payroll_accounts.find_by(pinwheel_account_id: argyle_account["id"])
    return unless payroll_account.present?

    # If we've made it here, there is a previous connection to that item.
    #
    # Redirect to the proper place based on the status of the connection.
    if payroll_account.sync_succeeded? || payroll_account.sync_failed?
      redirect_to cbv_flow_payment_details_path(user: { account_id: payroll_account.pinwheel_account_id })
    elsif payroll_account.job_succeeded?("accounts")
      # Sync is in progress, and the user has successfully connected their account.
      redirect_to cbv_flow_synchronizations_path(user: { account_id: payroll_account.pinwheel_account_id })
    end
  end

  def track_event
    return unless @cbv_flow.present?

    event_logger.track(TrackEvent::ApplicantBeganLinkingEmployer, request, {
      time: Time.now.to_i,
      cbv_flow_id: @cbv_flow.id,
      cbv_applicant_id: @cbv_flow.cbv_applicant_id,
      client_agency_id: @cbv_flow.client_agency_id,
      invitation_id: @cbv_flow.cbv_flow_invitation_id,
      item_id: params[:item_id],
      aggregator_name: "argyle"
    })
  end

  def argyle
    argyle_for(@cbv_flow)
  end
end
