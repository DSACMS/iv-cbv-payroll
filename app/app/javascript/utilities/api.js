import { fetchInternal } from "./fetchInternal"
import CSRF from "./csrf"

export const EVENTS_USER_ACTION = "/api/events/user_action"
export const PINWHEEL_TOKENS_GENERATE = "/api/pinwheel/tokens"
const ARGYLE_TOKENS_GENERATE = "/api/argyle/tokens"

export const trackUserAction = async (eventName, attributes = {}) => {
  return fetchInternal(EVENTS_USER_ACTION, {
    method: "post",
    body: JSON.stringify({ events: { event_name: eventName, attributes } }),
  })
}

export const fetchPinwheelToken = async (response_type, id, locale) => {
  const response = await fetch(PINWHEEL_TOKENS_GENERATE, {
    method: "post",
    headers: {
      "X-CSRF-Token": CSRF.token,
      "Content-Type": "application/json",
    },
    // Keep payload shape consistent with existing backend handling
    body: JSON.stringify({ response_type, id, locale }),
  })

  if (response.redirected) {
    window.location.href = response.url
    return
  }

  return response.json()
}

export const fetchArgyleToken = async (itemId) => {
  const response = await fetch(ARGYLE_TOKENS_GENERATE, {
    method: "post",
    headers: {
      "X-CSRF-Token": CSRF.token,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ item_id: itemId }),
  })

  if (response.redirected) {
    window.location.href = response.url
    return
  }

  return response.json()
}
