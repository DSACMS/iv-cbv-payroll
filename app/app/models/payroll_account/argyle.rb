class PayrollAccount::Argyle < PayrollAccount
  scope :awaiting_fully_synced_webhook, -> do
    joins(<<~SQL).where(webhook_events: { id: nil })
      LEFT OUTER JOIN webhook_events
      ON webhook_events.payroll_account_id = payroll_accounts.id
      AND webhook_events.event_name = 'users.fully_synced'
    SQL
  end

  def has_fully_synced?
    # Consider the sync finished if every webhook has returned either an error
    # or success. Any webhook with an "unknown" status will not impact the
    # result of this method.
    supported_jobs.all? do |job|
      supported_jobs.exclude?(job) ||
        find_webhook_event_for_job(job, "error").present? ||
        find_webhook_event_for_job(job, "success").present?
    end
  end

  def job_succeeded?(job)
    job_status(job) == :succeeded
  end

  def job_status(job)
    if supported_jobs.exclude?(job)
      :unsupported
    elsif find_webhook_event_for_job(job, "error").present?
      :failed
    elsif find_webhook_event_for_job(job, "success").present?
      :succeeded
    else
      :in_progress
    end
  end

  def necessary_jobs_succeeded?
    job_succeeded?("accounts") && (job_succeeded?("paystubs") || job_succeeded?("gigs"))
  end

  def find_webhook_event_for_job(job, event_outcome = nil)
    webhook_events.find do |webhook_event|
      Aggregators::Webhooks::Argyle::EVENT_NAMES_BY_JOB[job].include?(webhook_event.event_name) &&
        (event_outcome.nil? || webhook_event.event_outcome == event_outcome.to_s)
    end
  end

  def sync_started_at
    # Argyle sends `account.updated` webhook events while the user still has
    # the modal open (as they complete MFA). Since the user may take
    # arbitrarily long to complete MFA, we don't want to count this against the
    # sync time.
    account_connected_at = webhook_events
      .find { |webhook_event| webhook_event.event_name == "accounts.connected" }
      &.created_at

    # If, for some reason, we didn't get the `accounts.connected` event (?),
    # let's at least not crash anything that depends on this method.
    account_connected_at || created_at
  end

  def redact!
    argyle_environment = Rails.application.config.client_agencies[cbv_flow.client_agency_id].argyle_environment
    argyle = Aggregators::Sdk::ArgyleService.new(argyle_environment)
    argyle.delete_account_api(account: aggregator_account_id)
    touch(:redacted_at)
  rescue => ex
    raise ex unless Rails.env.production?

    Rails.logger.error "Unable to redact PayrollAccount::Argyle Account ID #{aggregator_account_id} - #{ex.message}"
    GenericEventTracker.new.track("DataRedactionFailure", nil, { account_id: aggregator_account_id })
  end
end
