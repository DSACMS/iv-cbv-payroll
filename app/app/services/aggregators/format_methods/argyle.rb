module Aggregators::FormatMethods::Argyle
  MILES_PER_KM = 0.62137

  # Note: this method is to map Argyle's employment status with Pinwheel's for consistency
  # between the two providers.
  def self.format_employment_status(employment_status)
    return unless employment_status

    case employment_status
    when "active"
      "employed"
    else
      employment_status
    end
  end

  def self.format_mileage(distance_string, distance_unit = "miles")
    return nil if distance_string.blank?
    distance = distance_string.to_f
    if distance_unit == "km"
      distance = distance * MILES_PER_KM
    end
    distance
  end

  def self.format_date(date)
    return unless date

    DateTime.parse(date).strftime("%Y-%m-%d")
  end

  def self.format_currency(amount)
    return unless amount
    dollars, cents = amount.split(".").map(&:to_i)

    (dollars * 100) + cents
  end

  def self.hours_computed(response_hours, response_gross_pay_list)
    if response_hours.present? && response_hours.to_f > 0
      response_hours.to_f
    else
      hours_by_earning_category(response_gross_pay_list).map { |_category, hours| hours.to_f }.max
    end
  end

  def self.hours_by_earning_category(gross_pay_list)
    gross_pay_list
       .filter { |e| e["hours"].present? }
       .group_by { |e| e["type"] }
       .transform_values { |earnings| earnings.sum { |e| e["hours"].to_f } }
  end

  def self.format_employer_address(a_paystub)
    return unless a_paystub.present? && a_paystub["employer_address"].present?
    employer_address = a_paystub["employer_address"]
    [
      employer_address["line1"],
      employer_address["line2"],
      "#{employer_address['city']}, #{employer_address['state']} #{employer_address['postal_code']}"
    ].compact.join(", ")
  end

  def self.seconds_to_hours(seconds)
    return unless seconds
    (seconds / 3600.0).round(2)
  end

  def self.employment_type(employment_type)
    if employment_type == "contractor"
      :gig
    else
      :w2
    end
  end

  def self.obfuscate_ssn(full_ssn)
    return unless full_ssn
    "XXX-XX-#{full_ssn.last(4).to_s.rjust(4, "X")}"
  end
end
