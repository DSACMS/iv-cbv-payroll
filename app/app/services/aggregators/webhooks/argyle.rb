module Aggregators::Webhooks
  class Argyle
    # When we receive a webhook from Argyle, we store the `status` for the
    # `job`(s) based on the webhook event name.
    #
    # E.g. "identites.added" indicates success of the identity and income jobs.
    #
    # For webhooks with `type: :partial`, additional logic must be applied to:
    # 1. Register the partially_synced webhooks (potentially multiple times)
    #    for any partial sync status needed for an agency.
    # 2. Determine whether the partially_synced success is sufficient to meet
    #    the agency's data requirements.
    #
    # For webhooks with `type: :include_resource, the webhook subscription will include the
    # { "include_resource": true } config to include the resource in the webhook payload.
    SUBSCRIBED_WEBHOOK_EVENTS = {
      "identities.added" => {
        status: :success,
        type: :non_partial,
        job: %w[identity income]
      },
      "paystubs.partially_synced" => {
        status: :success,
        type: :partial,
        job: %w[paystubs employment]
      },
      "gigs.partially_synced" => {
        status: :success,
        type: :partial,
        job: %w[gigs]
      },

      # We're not concerned with reporting these to the front-end/client.
      "accounts.connected" => {
        status: :success,
        type: :non_partial,
        job: []
      },
      "users.fully_synced" => {
        status: :success,
        type: :non_partial,
        job: %w[]
      },

      # The *.fully_synced events may take a long time to return; so we'll
      # store them for measurement purposes, but not tie any jobs' success to
      # their completion.
      "paystubs.fully_synced" => {
        status: :success,
        type: :non_partial,
        job: []
      },
      "gigs.fully_synced" => {
        status: :success,
        type: :non_partial,
        job: []
      },
      "accounts.updated" => {
        status: :success, # this gets overwritten dynamically on system_error in events_controller
        type: :include_resource,
        job: %w[accounts] # the accounts job is strictly used to pass state on system_error
      }
    }.freeze

    def self.generate_signature_digest(payload, webhook_secret)
      OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new("sha512"), webhook_secret, payload)
    end

    # Verify the signature using the appropriate webhook secret
    # See: https://docs.argyle.com/api-guide/webhooks
    def self.verify_signature(signature, payload, webhook_secret)
      expected = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new("sha512"), webhook_secret, payload)
      ActiveSupport::SecurityUtils.secure_compare(signature, expected)
    end

    def self.get_webhook_events(type: :non_partial)
      SUBSCRIBED_WEBHOOK_EVENTS.filter_map { |event_name, event| event_name if type == :all || event[:type] == type }
    end

    def self.get_supported_jobs
      SUBSCRIBED_WEBHOOK_EVENTS.values.map { |event| event[:job] }.flatten.compact.uniq
    end

    def self.get_webhook_event_outcome(event)
      SUBSCRIBED_WEBHOOK_EVENTS[event][:status]
    end

    def self.get_webhook_event_jobs(event)
      SUBSCRIBED_WEBHOOK_EVENTS[event][:job]
    end
  end
end
