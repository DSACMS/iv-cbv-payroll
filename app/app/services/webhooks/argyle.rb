module Webhooks
  class Argyle
    # Argyle's event outcomes are implied by the event name themselves i.e. accounts.failed (implies error)
    # x.fully_synced (implies success)
    # Define all webhook events we're interested in with their outcomes
    # and corresponding synchronizations page "job"
    SUBSCRIBED_WEBHOOK_EVENTS = {
      "users.fully_synced" => {
        status: :success,
        job: %w[income]
      },
      "identities.added" => {
        status: :success,
        job: %w[identity]
      },
      "paystubs.fully_synced" => {
        status: :success,
        job: %w[paystubs employment]
      },
      "gigs.fully_synced" => {
        status: :success,
        job: %w[gigs] # TODO: [FFS-XXX] update front-end/client to support gigs sync status
      },
      "accounts.connected" => {
        status: :success,
        job: [] # we're not concerned with reporting this to the front-end/client
      }
    }.freeze

    def self.generate_signature_digest(payload, webhook_secret)
      OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new("sha512"), webhook_secret, payload)
    end

    # Verify the signature using the appropriate webhook secret
    def self.verify_signature(signature, payload, webhook_secret)
      expected = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new("sha512"), webhook_secret, payload)
      ActiveSupport::SecurityUtils.secure_compare(signature, expected)
    end

    def self.get_webhook_events
      SUBSCRIBED_WEBHOOK_EVENTS.keys
    end

    def self.get_supported_jobs
      SUBSCRIBED_WEBHOOK_EVENTS.values.map { |event| event[:job] }.flatten.compact.uniq
    end

    def self.get_webhook_event_outcome(event)
      SUBSCRIBED_WEBHOOK_EVENTS[event][:status]
    end

    def self.get_webhook_event_jobs(event)
      SUBSCRIBED_WEBHOOK_EVENTS[event][:job]
    end
  end
end
