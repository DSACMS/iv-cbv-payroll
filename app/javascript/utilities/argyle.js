import loadScript from 'load-script';
import metaContent from "./meta";

const ARGYLE_TOKENS_REFRESH = '/api/argyle/tokens';

export function loadArgyle() {
  return new Promise((resolve, reject) => {
    loadScript('https://plugin.argyle.com/argyle.web.v5.js', (err, script) => {
      if (err) {
        reject(err);
      } else {
        resolve(Argyle);
      }
    });
  });
}

export function initializeArgyle(Argyle, userToken, callbacks) {
  return Argyle.create({
    userToken,
    sandbox: metaContent('argyle_sandbox'), // Set to false for production environment.
    ...callbacks
  });
}

export const updateToken = async updateToken => {
  const response = await fetch(ARGYLE_TOKENS_REFRESH).then(response => response.json());

  updateToken(response.token);
}
