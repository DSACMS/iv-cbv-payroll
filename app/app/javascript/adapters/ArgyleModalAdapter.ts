import { trackUserAction, fetchArgyleToken } from "@js/utilities/api.js";
import loadScript from 'load-script';
import { getDocumentLocale } from "@js/utilities/getDocumentLocale.js";
import { ModalAdapter } from "./ModalAdapter.js";

export default class ArgyleModalAdapter extends ModalAdapter {
  Argyle: Argyle;

  async load() {
    this.Argyle = await new Promise((resolve, reject) => {
      loadScript('https://plugin.argyle.com/argyle.web.v5.js', (err, script) => {
        if (err) {
          reject(err);
        } else {
          resolve(Argyle);
        }
      });
    })
  }

  async open() {
    const locale = getDocumentLocale();

    if (this.requestData) {
      await trackUserAction("ApplicantSelectedEmployerOrPlatformItem", {
        item_type: this.requestData.responseType,
        item_id: this.requestData.id,
        item_name: this.requestData.name,
        is_default_option: this.requestData.isDefaultOption,
        provider_name: this.requestData.providerName,
        locale
      })

      const { user } = await fetchArgyleToken();
      return this.Argyle.create({
        userToken: user.user_token,
        items: [this.requestData.id],
        onAccountConnected: this.onSuccess.bind(this),
        onTokenExpired: this.onTokenExpired.bind(this),
        onAccountCreated: async (payload) => { await trackUserAction("ArgyleAccountCreated", payload) },
        onAccountError: async (payload) => { await trackUserAction("ArgyleAccountError", payload) },
        onAccountRemoved: async (payload) => { await trackUserAction("ArgyleAccountRemoved", payload) },
        onClose: this.onClose.bind(this),
        onError: this.onError.bind(this),
        sandbox: true, 
      }).open();
    }
  }

  async onError(err: LinkError) {
    await trackUserAction("ArgyleError", err); 
    this.onExit()
  }

  async onClose() {
    await trackUserAction("ArgyleCloseModal"); 
    await this.onExit();
  }

  async onSuccess(eventPayload: ArgyleAccountData) {
    await trackUserAction("ArgyleSuccess", {
      account_id: eventPayload.accountId,
      user_id: eventPayload.userId,
      item_id: eventPayload.itemId,
    });

    if (this.successCallback) {
      this.successCallback(eventPayload.accountId);
    }
  }

  async onTokenExpired(updateToken : Function) {
      await trackUserAction("ArgyleTokenExpired")
      const { user } = await fetchArgyleToken();
      updateToken(user.user_token)
  }
  onEvent(eventName: string, eventPayload: any) {
  }
}
