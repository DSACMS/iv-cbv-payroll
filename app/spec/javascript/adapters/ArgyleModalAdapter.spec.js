import { vi, describe, beforeEach, afterEach, it, expect } from "vitest";
import loadScript from "load-script";
import ArgyleModalAdapter from "@js/adapters/ArgyleModalAdapter";
import { fetchArgyleToken, trackUserAction } from '@js/utilities/api';
import { mockArgyle, mockArgyleAuthToken } from "@test/fixtures/argyle.fixture";

const MOCK_ARGYLE_ERROR = "Failed to load SCRIPT"

const modalAdapterArgs = {
    onSuccess: vi.fn(),
    requestData: {
        responseType: "response-type",
        id: "id",
        providerName: "pinwheel",
        name: "test-name",
        isDefaultOption: true,
    }
}



describe('ArgyleModalAdapter', () => {
    let adapter;
    let triggers;
        
    beforeEach(async () => {
        mockArgyle();
        adapter = new ArgyleModalAdapter()
        adapter.load()
        adapter.init(modalAdapterArgs)
        triggers = await adapter.open()
    })
    afterEach(() => {
    })

    describe('open', () => {
        it('calls track user action', async () => {
            expect(trackUserAction).toHaveBeenCalled()
            expect(trackUserAction.mock.calls[0][0]).toBe('ApplicantSelectedEmployerOrPlatformItem')
            expect(trackUserAction.mock.calls[0]).toMatchSnapshot()
        })
        it('fetches token successfully', async () => {
            expect(fetchArgyleToken).toHaveBeenCalledTimes(1)
            expect(fetchArgyleToken).toHaveResolvedWith(mockArgyleAuthToken)
        })
        it('opens argyle modal', async () => {
            expect(Argyle.create).toHaveBeenCalledTimes(1)
        })
    })

    describe('Account Connected (success event)', () => {
        it('calls track user action', async() => {
            await triggers.triggerAccountConnected()
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('ArgyleSuccess')
        })
        it('triggers the modal adapter onSuccess callback', async () => {
            await triggers.triggerAccountConnected()
            expect(modalAdapterArgs.onSuccess).toHaveBeenCalled()
        })
    })
    describe.skip('onExit', () => {
        it('triggers the provided onExit callback', async () => {
            await triggers.triggerSuccessEvent()
            expect(pinwheelModalAdapterArgs.onSuccess).toHaveBeenCalled()
        })
    })
    describe.skip('onEvent', () => {
        it('logs screen_transition.LOGIN Event', async () => {
            await triggers.triggerEvent("screen_transition", { "screenName": "LOGIN", selectedEmployerName: "ACME Inc", selectedPlatformName: "ADP" })
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelShowLoginPage')
        })
        it('logs screen_transition.PROVIDER_CONFIRMATION event', async () => {
            await triggers.triggerEvent("screen_transition", { "screenName": "PROVIDER_CONFIRMATION"})
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelShowProviderConfirmationPage')
        })
        it('logs screen_transition.SEARCH_DEFAULT event', async () => {
            await triggers.triggerEvent("screen_transition", { "screenName": "SEARCH_DEFAULT"})
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelShowDefaultProviderSearch')
        })
        it('logs screen_transition.EXIT_CONFIRMATION event', async () => {
            await triggers.triggerEvent("screen_transition", { "screenName": "EXIT_CONFIRMATION"})
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelAttemptClose')
        })
        it('logs login_attempt event', async () => {
            await triggers.triggerEvent("login_attempt")
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelAttemptLogin')
        })
        it('logs error event', async () => {
            await triggers.triggerEvent("error", { type: "error-type", "code": "code", "message":"default message"})
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelError')
            expect(trackUserAction.mock.calls[1][1]['type']).toBe('error-type')
        })
        it('logs exit event', async () => {
            await triggers.triggerEvent("exit")
            expect(trackUserAction).toHaveBeenCalledTimes(2)
            expect(trackUserAction.mock.calls[1][0]).toBe('PinwheelCloseModal')
        })
    })
})