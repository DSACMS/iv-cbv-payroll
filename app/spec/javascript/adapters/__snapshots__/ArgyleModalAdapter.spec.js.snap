// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArgyleModalAdapter > open > calls track user action 1`] = `
[
  "ApplicantSelectedEmployerOrPlatformItem",
  {
    "is_default_option": true,
    "item_id": "id",
    "item_name": "test-name",
    "item_type": "response-type",
    "locale": "en",
    "provider_name": "pinwheel",
  },
]
`;
